core/types/receipt.go

```diff
@@ -82,8 +82,16 @@ type receiptRLP struct {
 	Logs              []*Log
 }
 
-// receiptStorageRLP is the storage encoding of a receipt.
-type receiptStorageRLP struct {
+// storedReceiptRLP is the storage encoding of a receipt.
+type storedReceiptRLP struct {
+	PostStateOrStatus []byte
+	CumulativeGasUsed uint64
+	ContractAddress   common.Address
```

Comment:
Can we also remove this field? Since we can derive the `contractAddress` if we assume the geth node should have the corresponding block body if we have `receipts`

Reply:
This is a good point! I was on the fence about this but my intuition was that most transactions do not create contracts so the overhead is just a byte (for the empty string RLP). Nonetheless, I think it's probably a good idea to just do it while we're here.

Reply:
Actually even an empty `contractAddress` is 20bytes.

Reply:
Ah, the other reason I decided not to do this was because it starts to become quite complicated. It would involve adding the following into `ReadReceipts` where the receipt metadata is populated:

```
if tx.To() == nil {
	genesisHash := ReadCanonicalHash(db, 0)
	config := ReadChainConfig(db, genesisHash)
	signer := types.MakeSigner(config, big.NewInt(int64(number)))
	from, err := types.Sender(signer, tx)
	receipt.ContractAddress = crypto.CreateAddress(from, tx.Nonce())
}
```

So this would introduce two additional reads (on top of the body read) and incur the overhead of extracting the sender from the transaction.

I'm fine to proceed however you and @karalabe would like. IMHO I would probably go for computing the contract address just-in-time for a read and incur these overheads. Adopting #19200 would allow us to cache the `ChainConfig` so that lookup can be removed.

Should we benchmark? I'm going to add this code in for now.

Reply:
> Actually even an empty contractAddress is 20bytes.
My bad. Did not see that `Address` is  `[AddressLength]byte`. Looks like a good win then!

Reply:
@rjl493456442 After thinking about this a bit there are some subtleties to `contract_address`:
1. The transaction has to succeed for the contract address to be there
2. There is no way to tell if a pre-Byzantium contract succeeds

That means we probably should store `contract_address` pre-Byzantium for contract creation transactions (or even for any contract transaction?). We can certainly omit storing this for call transactions, which should still give us a pretty big win. 

@rjl493456442 @karalabe What are your thoughts?

Reply:
@Matthalp The `contract_address` should always be there no matter the transaction execution is successful or not. Please check the relative code https://github.com/ethereum/go-ethereum/blob/master/core/state_transition.go#L183,
a transaction will be discarded only if it doesn't pass the precheck or doesn't have enough balance to make the first transfer. This type of transaction should be included in the block otherwise there will cause a consensus error.
All other transactions will have the receipt (with the contract_address field if it is a creation tx. :)))

Reply:
I was more or less pointing out that the `contract_address` field can sometimes be invalid since a failed `CREATE` transaction would not actually result in creating `contract_address` (the code is not set and funds are not deposited). Its presence under these conditions is probably outside the scope of this PR, but figured I would raise it just in case.

FWIW I implemented block processing for all pre-Constantinople hard forks in the Pantheon Ethereum client, so I'm fairly familiar with the semantics there. I am less familiar with JSON-RPC specifications, which is where this issue seems to stems from.

Reply:
Sure, I totally agree with your point.